### PR TITLE
AI MAGIC!

### PR DESCRIPTION
**Pull request by NegativeScript**
More to come inshaAllah.

### CODE REVIEW

File: furusiya/function_mess.py [hunk 1]
```diff
@@ -122,7 +122,7 @@ def process_bow():
                 elif event.type == 'KEYDOWN':
                     if event.key == 'ESCAPE':
                         Game.draw_bowsight = False
-                        break
+                        return 'didnt-take-turn'
```
Comment (nightblade9) [inline furusiya/function_mess.py]: This is now a constant. We shouldn't be using strings.

Please double-check that this doesn't break if you:
- Press `f`
- Press escape

File: furusiya/model/components/ai/base.py [hunk 2]
```diff
@@ -1,10 +1,27 @@
...
-    def take_turn(self):
+    def __init__(self, owner, num_turns=None):
+        super().__init__(owner)
+        self.num_turns = num_turns
+        self.take_turn = self._take_turn
+
+    def _take_turn(self):
         raise NotImplementedError()
+
+    def set_temporary(self, other):
```
Comment (nightblade9) [inline furusiya/model/components/ai/base.py]: This method needs a better name. Maybe `temporarily_switch_to`?

File: furusiya/model/components/ai/base.py [hunk 3]
```diff
@@ -1,10 +1,27 @@
...
+        AISystem.set_ai(self.owner, other)
+
+        def new_take_turn():
```
Comment (nightblade9) [inline furusiya/model/components/ai/base.py]: This definitely needs a better name. Maybe `process_temporary_ai_turn`?

File: furusiya/model/components/ai/base.py [hunk 4]
```diff
@@ -1,10 +1,27 @@
 from model.components.base import Component
+from model.systems.ai_system import AISystem
 
 
 class AI(Component):
```
Comment (nightblade9) [inline furusiya/model/components/ai/base.py]: This should really be called `BaseAI` or `AbstractAI` or something

File: furusiya/function_mess.py (hunk 1 above)
Comment (NegativeScript) [reply furusiya/function_mess.py]: Nope. Works as expected.

However, that is a good point. I'll open an issue to convert constant-ish strings into enums.

File: tests/furusiya/model/systems/test_ai_system.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    ai_sys.set_ai(owner3, ai3)
+
+    ai_sys.take_monster_turns()
+
+    for ai in (ai1, ai2, ai3):
+        assert ai.take_turn.called
+
+    ai_sys.remove_ai(owner1)
+    assert not ai_sys.has_ai(owner1)
+    assert ai_sys.get_ai(owner1) is None
```
Comment (nightblade9) [inline tests/furusiya/model/systems/test_ai_system.py]: Can you add an assert that `ai1.take_turn` is not called?

**Pull request merged**